### ISSUE TITLE
PNG write error macM1

### ISSUE THREAD

**Issue by wtobin** [bf777/MesoNet#18]
Hi,

I am trying to get MesoNet running on a mac M1 laptop. I installed DeepCut using miniconda3 and MesoNet according to the instructions on the github main page. When I run 'test.py' it fails with the following output:

% python test.py 
Loading DLC 2.3.5...
Successfully created the directory /Users/wtobin/MesoNet/mesonet/../tests/results/mesonet_output_atlas_brain 
Successfully created the directory /Users/wtobin/MesoNet/mesonet/../tests/results/mesonet_output_brain_atlas 
Successfully created the directory /Users/wtobin/MesoNet/mesonet/../tests/results/mesonet_output_sensory 
Successfully created the directory /Users/wtobin/MesoNet/mesonet/../tests/results/mesonet_output_MBFM_U_Net 
Successfully created the directory /Users/wtobin/MesoNet/mesonet/../tests/results/mesonet_output_voxelmorph 
/Users/wtobin/MesoNet/mesonet/models/DongshengXiao_brain_bundary.hdf5
1. Atlas-to-brain warp with U-Net and DeepLabCut
/Users/wtobin/MesoNet/mesonet
/Users/wtobin/MesoNet/mesonet/models/DongshengXiao_brain_bundary.hdf5
2023-07-05 19:30:30.923781: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
2023-07-05 19:30:30.946077: W tensorflow/c/c_api.cc:300] Operation '{name:'conv2d_24/bias/Assign' id:601 op device:{requested: '', assigned: ''} def:{{{node conv2d_24/bias/Assign}} = AssignVariableOp[_has_manual_control_dependencies=true, dtype=DT_FLOAT, validate_shape=false](conv2d_24/bias, conv2d_24/bias/Initializer/zeros)}}' was changed by setting attribute after it was run by a session. This mutation will have no effect, and will trigger an error in the future. Either don't modify nodes after running them or create a new session.
2023-07-05 19:30:32.398229: W tensorflow/c/c_api.cc:300] Operation '{name:'conv2d_21/bias/v/Assign' id:1423 op device:{requested: '', assigned: ''} def:{{{node conv2d_21/bias/v/Assign}} = AssignVariableOp[_has_manual_control_dependencies=true, dtype=DT_FLOAT, validate_shape=false](conv2d_21/bias/v, conv2d_21/bias/v/Initializer/zeros)}}' was changed by setting attribute after it was run by a session. This mutation will have no effect, and will trigger an error in the future. Either don't modify nodes after running them or create a new session.
/Users/wtobin/MesoNet/mesonet/../tests/sample_data/pipeline1_2
/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/mesonet-1.0.6.1-py3.9.egg/mesonet/predict_regions.py:93: UserWarning: `Model.predict_generator` is deprecated and will be removed in a future version. Please use `Model.predict`, which supports generators.
/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/keras/engine/training_v1.py:2359: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  updates=self.state_updates,
2023-07-05 19:30:33.191665: W tensorflow/c/c_api.cc:300] Operation '{name:'conv2d_24/Sigmoid' id:607 op device:{requested: '', assigned: ''} def:{{{node conv2d_24/Sigmoid}} = Sigmoid[T=DT_FLOAT, _has_manual_control_dependencies=true](conv2d_24/BiasAdd)}}' was changed by setting attribute after it was run by a session. This mutation will have no effect, and will trigger an error in the future. Either don't modify nodes after running them or create a new session.
Traceback (most recent call last):
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/PIL/PngImagePlugin.py", line 1277, in _save
    rawmode, mode = _OUTMODES[mode]
KeyError: 'F'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/wtobin/MesoNet/tests/test.py", line 92, in <module>
    mesonet.predict_regions(config_file_atlas_brain)
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/mesonet-1.0.6.1-py3.9.egg/mesonet/predict_regions.py", line 153, in predict_regions
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/mesonet-1.0.6.1-py3.9.egg/mesonet/predict_regions.py", line 95, in predictRegion
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/mesonet-1.0.6.1-py3.9.egg/mesonet/mask_functions.py", line 116, in saveResult
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/skimage/io/_io.py", line 143, in imsave
    return call_plugin('imsave', fname, arr, plugin=plugin, **plugin_args)
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/skimage/io/manage_plugins.py", line 205, in call_plugin
    return func(*args, **kwargs)
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/imageio/v3.py", line 147, in imwrite
    encoded = img_file.write(image, **kwargs)
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/imageio/core/v3_plugin_api.py", line 367, in __exit__
    self.close()
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/imageio/plugins/pillow.py", line 123, in close
    self._flush_writer()
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/imageio/plugins/pillow.py", line 457, in _flush_writer
    primary_image.save(self._request.get_file(), **self.save_args)
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/PIL/Image.py", line 2413, in save
    save_handler(self, fp, filename)
  File "/Users/wtobin/miniconda/envs/DEEPLABCUT_M1/lib/python3.9/site-packages/PIL/PngImagePlugin.py", line 1280, in _save
    raise OSError(msg) from e
OSError: cannot write mode F as PNG

Hoping you can help me sort this out,

-Willie

**Comment by wtobin** [comment]
I have now encountered the same error on two other machines, an intel mac running Ventura and Linux machine running Ubuntu 20.04

**Comment by wtobin** [comment]
I think this occurs because the code attempts to save floating point number arrays as png images, which is not permitted. I would love some clarification as to why it was written this way, what I am missing.

**Comment by bf777** [comment]
Hi @wtobin, thanks for letting me know about this issue! It was indeed related to images being saved as floats instead of uint8 (which I've now corrected). I've just pushed a change that should fix this.

**Comment by bf777** [comment]
I've updated the code a bit more, the culprit for the latest issue may have been the neurite package and its implementation of tensorflow - that package isn't necessary to use right now. I've kept the keras and protobuf versions as they are for now as newer versions of tensorflow on different systems (including mine) bump up the requirements for these. If you run into any keras or protobuf-specific errors, let me know!
